torch/_logging/internal.py

```diff
@@ -0,0 +1,384 @@
+        log.removeHandler(handler)
+
+def _get_registered_parent_qname(log_qname):
+    logger = logging.getLogger(log_qname)
+    while logger.parent:
+        if log_registry.is_log(logger.parent):
+            return logger.parent.name
+        logger = logger.parent
+
+    return None
+
+# initialize loggers log_names
+# each developer component should call this for their own logs
+# in the appropriate location after relevant types have been registered
+def _init_logs(log_file_name=None):
+    for log_name in log_registry.get_log_qnames():
+        log = logging.getLogger(log_name)
+        log.setLevel(logging.DEBUG)  # allow all messages through to the handlers
+        log.propagate = False
+        _clear_handlers(log)
+
+    log_state = _get_log_state()
+    log_qname_to_enabled_types: DefaultDict[str, Set[type]] = collections.defaultdict(set)
+    log_qname_to_level = dict()
+
+    # generate a map of log_name -> the types that should be logged
+    for name, level in log_state.log_name_to_level.items():
+        log_qname = log_registry.name_to_log_qname[name]
+        assert log_registry.is_log(name)
+        log_qname_to_level[log_qname] = level
+        logging.getLogger(log_qname).setLevel(
+            logging.DEBUG
+        )  # allow all messages through logger
+        # ensure log_name is in the dictionary
+        rec_types = log_qname_to_enabled_types[log_qname]
+        if level == logging.DEBUG or not log_registry.supports_verbosity(name):
+            rec_types.update(log_registry.log_qname_to_rec_types[log_qname])
+
+    for name in log_state.enabled_artifact_names:
+        log_qname = log_registry.name_to_log_qname[name]
+        log_qname_to_enabled_types[log_qname].add(log_registry.name_to_rec_type[name])
+
+    for log_name, enabled_types in log_qname_to_enabled_types.items():
+        log = logging.getLogger(log_name)
+        level = log_qname_to_level.get(log_name, DEFAULT_LOG_LEVEL)
+
+        _setup_handlers(
+            lambda: _tag_handler(logging.StreamHandler()),
+            log,
+            enabled_types,
+            level,
+        )
+
+        if log_file_name is not None:
+            _setup_handlers(
+                lambda: _tag_handler(logging.FileHandler(log_file_name)),
+                log,
+                enabled_types,
+                level,
+            )
+
+
+@functools.lru_cache(None)
+def warning_once(self, *args, **kwargs):
+    """
+    This method is identical to `logger.warning()`, but will emit the warning with the same message only once
+    Note: The cache is for the function arguments, so 2 different callers using the same arguments will hit the cache.
+    The assumption here is that all warning messages are unique across the code. If they aren't then need to switch to
+    another type of cache that includes the caller frame information in the hashing function.
+    """
+    self.warning(*args, **kwargs)
+
+
+logging.Logger.warning_once = warning_once
```

Comment:
Monkey patching into stdlib just so that you can write `log.warning_once` instead of something a little longer is bad form. Just make people call `torch._warning_once` or something. (Or make all of our loggers subclasses? Your choice.)

Reply:
I'll just make torch._warning_once, I think this is fine. Personally I actually think the monkey patching is a better ux in this case, but it's marginal, so I'll change this. 

There is a way to extend the logger class - but this extends it for all loggers, including ones we don't own, and basically the last call to setting the logger class wins, so this is bad for interoperability.

Reply:
That's probably I did it in the first place, it's been a long time.

The other approach would be to extend not the Logger class, but the logger object itself. In which case the scope is very controlled and won't impact any other uses of `logging.Logger` in the composite system of pytorch is only a part.

Reply:
This would be trickier I think, since you'd have to change all call sites of getLogger() to monkey patch the object, right?

Reply:
Considering all of the options, the one line patch of the Logger seems most acceptable to me, all of the others have much larger issues. I'm going to keep this, let me know if either of you feel really strongly about this, and I can just do torch._warning_once